content/learning-paths/mobile-graphics-and-gaming/android_halide/processing-workflow.md

```diff
@@ -0,0 +1,585 @@
+static inline Halide::Expr clampCoord(Halide::Expr coord, int maxCoord) {
+    return Halide::clamp(coord, 0, maxCoord - 1);
+}
+
+int main() {
+    // Open the default camera.
+    VideoCapture cap(0);
+    if (!cap.isOpened()) {
+        cerr << "Error: Unable to open camera." << endl;
+        return -1;
+    }
+
+    while (true) {
+        // Capture frame (typically interleaved BGR).
+        Mat frame;
+        cap >> frame;
+        if (frame.empty()) {
+            cerr << "Error: Received empty frame." << endl;
+            break;
+        }
+        if (!frame.isContinuous()) frame = frame.clone();
+
+        const int width    = frame.cols;
+        const int height   = frame.rows;
+        const int channels = frame.channels(); // 3 (BGR) or 4 (BGRA)
+
+        // Wrap the interleaved OpenCV frame for Halide.
+        auto in_rt = Halide::Runtime::Buffer<uint8_t>::make_interleaved(
+            frame.data, width, height, channels);
+        Halide::Buffer<> inputBuffer(*in_rt.raw_buffer()); // front-end view
+
+        // Define ImageParam (x, y, c) and declare interleaved layout.
+        Halide::ImageParam input(Halide::UInt(8), 3, "input");
+        input.set(inputBuffer);
+        input.dim(0).set_stride(channels); // x-stride = C (interleaved)
+        input.dim(2).set_stride(1);        // c-stride = 1 (adjacent bytes)
+        input.dim(2).set_bounds(0, channels);
+
+        // Spatial vars.
+        Halide::Var x("x"), y("y");
+
+        // Grayscale in Halide
+        Halide::Func gray("gray");
+        Halide::Expr r16 = Halide::cast<int16_t>(input(x, y, 2));
+        Halide::Expr g16 = Halide::cast<int16_t>(input(x, y, 1));
+        Halide::Expr b16 = Halide::cast<int16_t>(input(x, y, 0));
+
+        // Integer approx: Y ≈ (77*R + 150*G + 29*B) >> 8
+        gray(x, y) = Halide::cast<uint8_t>((77 * r16 + 150 * g16 + 29 * b16) >> 8);
+
+        // 3×3 binomial kernel (sum = 16).
+        int kernel_vals[3][3] = {
+            {1, 2, 1},
+            {2, 4, 2},
+            {1, 2, 1}
+        };
+        Halide::Buffer<int> kernelBuf(&kernel_vals[0][0], 3, 3);
+
+        // Blur via reduction over a 3×3 neighborhood.
+        Halide::RDom r(0, 3, 0, 3);
+        Halide::Func blur("blur");
+
+        // Use int16_t for safe multiply-and-accumulate with 8-bit input.
+        Halide::Expr val =
+            Halide::cast<int16_t>(
+                gray(clampCoord(x + r.x - 1, width),
+                     clampCoord(y + r.y - 1, height))
+            ) * Halide::cast<int16_t>(kernelBuf(r.x, r.y));
+
+        blur(x, y) = Halide::cast<uint8_t>(Halide::sum(val) / 16);
+
+        // Thresholding.
+        Halide::Func thresholded("thresholded");
+        thresholded(x, y) = Halide::cast<uint8_t>(
+            Halide::select(blur(x, y) > 128, 255, 0)
+        );
+
+        // Realize and display.
+        Halide::Buffer<uint8_t> outputBuffer;
+        try {
+            outputBuffer = thresholded.realize({ width, height });
```

Comment:
We should move pipeline construction code outside of iteration loop. Otherwise, compilation is performed for each iteration, which significantly slows performance and makes the effect of scheduling invisible.